peps/pep-0810.rst

```diff
@@ -0,0 +1,1345 @@
+3. Test that side-effect timing changes don't break functionality
+4. Use ``__lazy_modules__`` for compatibility with older Python versions
+
+**Q: What about star imports** (``from module import *``)?
+
+A: Wild card (star) imports cannot be lazy - they remain eager. This is because the
+set of names being imported cannot be determined without loading the module. Using the
+``lazy`` keyword with star imports will be a syntax error. If lazy imports are globally
+enabled, star imports will still be eager.
+
+**Q: How do lazy imports interact with import hooks and custom loaders?**
+
+A: Import hooks and loaders work normally. When a lazy object is first used, the
+standard import protocol runs, including any custom hooks or loaders that were
+in place at reification time.
+
+**Q: What happens in multi-threaded environments?**
+
+A: Lazy import reification is thread-safe. Only one thread will perform the
+actual import, and the binding is atomically updated. Other threads will see
+either the lazy proxy or the final resolved object.
+
+**Q: Can I force reification of a lazy import without using it?**
+
+A: Yes, accessing a module's ``__dict__`` will reify all lazy objects in that
+module. Individual lazy objects can be resolved by calling their ``get()`` method.
+
+**Q: What's the difference between** ``globals()`` **and** ``mod.__dict__`` **for lazy imports?**
+
+A: Calling ``globals()`` returns the module's dictionary without reifying lazy
+imports — you'll see lazy proxy objects when accessing them through the returned
+dictionary. However, accessing ``mod.__dict__`` from external code reifies all lazy
+imports in that module first. This design ensures:
+
+.. code-block:: python
+
+  # In your module:
+  lazy import json
+
+  g = globals()
+  print(type(g['json']))  # <class 'lazy_import'> - your problem
+
+  # From external code:
+  import sys
+  mod = sys.modules['your_module']
+  d = mod.__dict__
+  print(type(d['json']))  # <class 'module'> - reified for external access
+
+This distinction means adding lazy imports and calling ``globals()`` is your
+responsibility to manage, while external code accessing ``mod.__dict__`` always
+sees fully loaded modules.
+
+**Q: Why not use** ``importlib.util.LazyLoader`` **instead?**
+
+A: ``LazyLoader`` has significant limitations:
+
+- Requires verbose setup code for each lazy import
+- Has ongoing performance overhead on every attribute access
+- Doesn't work well with ``from ... import`` statements
+- Less clear and standard than dedicated syntax
+
+**Q: Will this break tools like** ``isort`` **or** ``black``?
+
+A: Tools will need updates to recognize the ``lazy`` keyword, but the changes
+should be minimal since the import structure remains the same. The keyword
+appears at the beginning, making it easy to parse.
+
+**Q: How do I know if a library is compatible with lazy imports?**
+
+A: Most libraries should work fine with lazy imports. Libraries that might have issues:
```

Comment:
This seems most of an issue with the flag that enables lazy imports globally, for what it's worth. If you just put lazy imports in your code using the library, it's very difficult for it to break a third-party library.

Reply:
That's unfortunately not true, it's a potential problem even with explicit lazy imports.